streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -582,7 +596,7 @@ public Subscription subscription(final Set<String> topics) {
 
         // construct the global partition assignment per host map
         final Map<HostInfo, Set<TopicPartition>> partitionsByHostState = new HashMap<>();
-        if (minReceivedMetadataVersion == 2 || minReceivedMetadataVersion == 3) {
+        if (minReceivedMetadataVersion >= 2) {
```

Comment:
This is another meta comment, not related to this line: in `CopartitionedTopicsValidator` we should also update the logic accordingly, first of the

```
if (partitions == null) {
                        throw new org.apache.kafka.streams.errors.TopologyException(String.format("%sTopic not found: %s", logPrefix, topic));
                    }
```

Should never happen, since we would already fail before if the metadata is not complete, and the `NOT_AVAILABLE` case should not happen either (see my other comment).

Also note that there is a related bug fix PR for this JIRA long time ago about when the ensureCopartitioning should be called: https://github.com/apache/kafka/pull/2815/files, with this general change:

```
if (numPartitions == UNKNOWN) {
                for (final Map.Entry<String, InternalTopicMetadata> entry: allRepartitionTopicsNumPartitions.entrySet()) {
                    if (copartitionGroup.contains(entry.getKey())) {
                        final int partitions = entry.getValue().numPartitions;
                        if (partitions > numPartitions) {
                            numPartitions = partitions;
                        }
                    }
                }
            }
```

Should not happen either since all topic's num.partitions should be determined by then.

Reply:
Again this is another meta comment: the member decoding and handling leader's propagated assignment is in `onAssignment`, in which the we decode `AssignmentInfo` from `Assignment#userData`. In that function we should check the returned error code in `AssignmentInfo`, and if it is not NONE we should "gracefully" shutdown than just throwing a runtime exception: for example, we can set a flag indicating we need to error out, and then in `onPartitionsAssigned` callback we can check this flag and then decide to shutdown if necessary.

Reply:
For onPartitionsAssigned, did you mean the method in StreamThread ?
The method takes Collection<TopicPartition>. Does this mean the flag should be added to TopicPartition ?